### PR TITLE
ARROW-5264: [Java] Allow enabling/disabling boundary checking by environmental variable

### PR DESCRIPTION
**Pull request by liyafan82**
The flag BoundsChecking#BOUNDS_CHECKING_ENABLED determines if boundary checking is enabled/disabled in vector/arrow buffer APIs. 
It has significant performance implications, since boundary checking is a frequent operation.

This PR addresses 2 problems with the flag for boundary checking in Java API:

1. The flag can be determined by an environmental variable: ARROW_ENABLE_UNSAFE_MEMORY_ACCESS, in addition to the system properties. The system properties have higher priority than the environmental variable.

2. There is an old and a new system property for this flag. To disable boundary checking, both the old and new properties must be set to true, which is undesirable. In addition, if both old and new properties are provided, the behavior is indeterminate. In the patch, the new property takes precedence over the old property.

### CODE REVIEW

**Comment by jacques-n** [comment]
-1 
This was a design choice to maximize the likelihood that the choice is never actually evaluated (since it is either always true or always false for all vectors across time).

**Comment by liyafan82** [comment]
> -1
> This was a design choice to maximize the likelihood that the choice is never actually evaluated (since it is either always true or always false for all vectors across time).

I see. Thanks. In the new commit, this design choice is left unaffected. I only add an environmental variable to set the flag. The environmental variable has a lower priority compared to the system properties.

**Comment by jacques-n** [comment]
Why are you removing tests in this commit?

**Comment by liyafan82** [comment]
> Why are you removing tests in this commit?

Added them back. Thanks for your kind reminder.

**Comment by liyafan82** [comment]
Hi @jacques-n, do you have any more comments? Do you have any suggestions to the test cases?

File: java/memory/src/main/java/org/apache/arrow/memory/BoundsChecking.java [hunk 1]
```diff
@@ -30,17 +30,21 @@
...
-    assert isAssertEnabled = true;
-    BOUNDS_CHECKING_ENABLED = isAssertEnabled ||
-      !"true".equals(System.getProperty("arrow.enable_unsafe_memory_access")) ||
-      !"true".equals(oldProperty);
+    String newProperty = System.getProperty("arrow.enable_unsafe_memory_access");
+
+    // The priority of determining the unsafe flag:
+    // 1. The system properties take precedence over the environmental variable.
+    // 2. The new system property takes precedence over the new system property.
+    String unsafeFlagValue = newProperty != null ? newProperty : (oldProperty != null ? oldProperty : envProperty);
```
Comment (jacques-n) [inline java/memory/src/main/java/org/apache/arrow/memory/BoundsChecking.java]: stacking ternary operators is quite hard to read, suggest against it in general.

**Comment by jacques-n** [comment]
+1 lgtm

@liyafan82 , thanks for the contribution. I suggest you add a similar flag for isSet() to be optionally skipped in a separate jira to improve performance

File: java/memory/src/main/java/org/apache/arrow/memory/BoundsChecking.java (hunk 1 above)
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/BoundsChecking.java]: Thanks for the suggestion. I have replace them with if conditions.

**Comment by liyafan82** [comment]
> +1 lgtm
> 
> @liyafan82 , thanks for the contribution. I suggest you add a similar flag for isSet() to be optionally skipped in a separate jira to improve performance

Thanks a lot for your suggestion. I will start a separate JIRA for that.

**Comment by emkornfield** [comment]
Please update javadoc

**Comment by liyafan82** [comment]
> Please update javadoc

@emkornfield updated. Thanks for your kind reminder.

**Comment by emkornfield** [comment]
There is class level docs that need updating also I think

**Comment by liyafan82** [comment]
> There is class level docs that need updating also I think

@emkornfield I have updated the javadoc for class org.apache.arrow.memory.BoundsChecking. Thanks a lot for your careful review.

**Comment by emkornfield** [comment]
+1 FYI, I also think this fixes https://issues.apache.org/jira/browse/ARROW-4851, sorry not at my development computer or I would merge

**Comment by liyafan82** [comment]
> deprecated.

Resolved.
Thanks again for the careful review, and sorry for my poor English.

**Comment by liyafan82** [comment]
> +1 FYI, I also think this fixes https://issues.apache.org/jira/browse/ARROW-4851, sorry not at my development computer or I would merge (after typo is fixed)

Hi @emkornfield , I have fixed the typo. Would you please help merge it?

**Comment by wesm** [comment]
Merging this

**Pull request closed without merging**